exo/networking/manual/manual_discovery.py

```diff
@@ -0,0 +1,174 @@
+import asyncio
+import time
+import traceback
+from typing import List, Dict, Callable, Tuple
+from exo.networking.discovery import Discovery
+from exo.networking.peer_handle import PeerHandle
+from exo.topology.device_capabilities import DeviceCapabilities, DeviceFlops
+from exo.helpers import DEBUG, DEBUG_DISCOVERY
+from exo.networking.grpc.grpc_peer_handle import GRPCPeerHandle
+import yaml
+import socket
+
+class ManualDiscovery(Discovery):
+  def __init__(
+    self,
+    create_peer_handle: Callable[[str, str, DeviceCapabilities], PeerHandle],
+    discovery_config: str = "topology.yml",
+    discovery_interval: int = 5,
+    discovery_timeout: int = 30,
+    update_interval: int = 15,
+  ):
+    self.discovery_config = discovery_config
+    with open(self.discovery_config, 'r') as f:
+      config_devices = yaml.safe_load(f)
+      f.close()
+
+    for device in config_devices:
+      print(f"{device['server']} {device['id']}:")
+      print(f"I am {socket.gethostname()} !")
+      if f"{socket.gethostname()}" == f"{device['server']}":
+        print(f"Adresse: {device['address']} {device['port']}")
+        self.node_id = f"{device['id']}"
+        self.node_address = f"{device['address']}"
+        self.node_port = f"{device['port']}"
+        print(f"Capabilities:")
+        for capability, value in device['device_capabilities'].items():
+          print(f"{capability}: {value}")
+          if f"{capability}" == "model":
+            self.model = f"{value}"
+          if f"{capability}" == "chip":
+            self.chip = f"{value}"
+          if f"{capability}" == "memory":
+            self.memory = (int(f"{value}"))
+          if f"{capability}" == "fp32":
+            self.fp32 = (int(f"{value}"))
+          if f"{capability}" == "fp16":
+            self.fp16 = (int(f"{value}"))
+          if f"{capability}" == "int8":
+            self.int8 = (int(f"{value}"))
+    
+    self.create_peer_handle = create_peer_handle
+    self.discovery_interval = discovery_interval
+    self.discovery_timeout = discovery_timeout
+    self.update_interval = update_interval
+    self.device_capabilities=GRPCPeerHandle(self.node_id, self.node_address, DeviceCapabilities(model=self.model, chip=self.chip, memory=self.memory, flops=DeviceFlops(fp32=self.fp32, fp16=self.fp16, int8=self.int8)))
+    self.known_peers: Dict[str, Tuple[PeerHandle, float, float]] = {}
+    self.discovery_task = None
+    self.cleanup_task = None
+    self._device_id = None
+    self.update_task = None
+
+  def get_device_capabilities(self, gethostname) -> DeviceCapabilities:
```

Comment:
Same, get_device_capabilities is just device_capabilities I think, but I wanted to avoid a conflict.
Move config_devices okay but where do you want to manage the read for the manual file ??

Reply:
Move the function outside of the class. If there's a name conflict, name it something else

Reply:
Hello, just put the read of topology in ReadManualConfig class.